Remote Data Paradigm

evandana opened the pull request:
# Summary
Allows grid to accommodate server-centric data source

## Example
- Given: 10,000 alarms in a system
- UI would request one page's worth of data, 10 records
- UI would be provided with meta data such as total number of records (10,000)
- UI can display one page while generating pagination UI for all 10,000

# Technical Debt
- `demo.html` page been added, but could use review (separate effort from this PR)
  - Work on demo page done by [@dcheepurupalli](https://github.com/dcheepurupalli)
  - Help is welcomed

# Technical Notes
```
<px-data-table 
            data-remote="true"
            enable-column-resize="true"
            filterable$={{_filterable}} 
            first-item-index="{{_firstItemIndex}}"
            greedy-height-with-scroll="true"
            id="alarmGrid" 
            page-size="{{_pageSize}}" 
            show-column-chooser="true" 
            table-data="[[_alarms]]" 
            table-rows="true" 
            total-entries="{{_filteredCount.filtered}}"
            >
</px-data-table>
```

mdwragg commented:
Awesome, thanks @evandana! First things first, I note some unit tests are failing, can we please get those working and then we've got a base on which to make review changes.

Many thanks for your hard work, we really appreciate it!

evandana commented:
@mdwragg `wct` tests pass now. Let me know if you have any further questions.

evandana commented:
Merged in upstream changes that occurred while PR has been open

mdwragg commented:
Hi,

@evandana can you revert the formatting you have applied to the source files? I'm noticing 4 spacing. This shouldn't happen if you have the editorconfig plugin active and working (and picking up our `.editorconfig` settings.

It's also going to potentially make merging more error prone/difficult. Thanks!

Also, I pulled the latest to the PR and the review comments I made above ^^ haven't been addressed (e.g. I see the `coverage` folder still in your branch.

Thanks!

evandana commented:
@mdwragg: I've combed through the template on aha-table.html to reduce whitespace diffs. Note that there are aesthetic changes in my code for sorting attributes alphabetically and grouping methods by function. To view the diff [ignoring whitespace changes](https://github.com/blog/967-github-secrets#user-content-whitespace), append `?w=1` to the URL, e.g. https://github.com/PredixDev/px-data-table/pull/126/files?w=1.

In your recent comment you hinted at previous comments you made regarding the `coverage` folder, etc. I haven't seen any of these comments and would greatly appreciate your help locating them.

mdwragg commented:
@evandana - apologies, I think I needed to hit the 'submit review button', I could see the review comments but I don't think you could, sorry about that!

evandana commented on sass/aha-table-predix.scss:
```diff
@@ -20,10 +20,52 @@ common/abstract rules go in px-data-table-sketch.scss, not in this file.
+  }
+  /* You could think of as "step 2" */
+  20% {
+    background: cyan;
```
Removing this

evandana commented on sass/aha-table-predix.scss:
```diff
@@ -20,10 +20,52 @@ common/abstract rules go in px-data-table-sketch.scss, not in this file.
   background-color: var(--px-data-table-background-color-striped--even, $gray2);
 }
 
+:host[greedy-height-with-scroll] {
```
`greedy-height-with-scroll` will enable the table to expand in height to fill container

While the `greedy-height` functionality could be considered separate, it has been quite integrated into this codebase. It only adds a little bit of styling. Would it be possible to have it reviewed with this same PR?

randyaskin replied:
In that case, refer to the comments I made on Tina's PR and the subsequent changes she made - https://github.com/PredixDev/px-data-table/pull/117#issuecomment-292346117 . Also, I think all of the tests for greedy-height were removed at some point?

evandana replied:
Thanks @randyaskin. I'll either update by removing "greedyHeight" or modifying it per your comments.

evandana commented:
- Merged in latest from `master`
- Ensured that all unit tests pass

@mdwragg @randyaskin

mdwragg commented:
Hi, I couldn't successfully recompile the sass when I looked at this PR locally...

Also, the checked in CSS broke the filter header in both the complex demo and the demo.html page.

evandana commented:
I'll look at that momentarily.
On Thu, Jun 1, 2017 at 7:49 PM Martin Wragg <notifications@github.com>
wrote:

> Hi, I couldn't successfully recompile the sass when I looked at this PR
> locally...
>
> Also, the checked in CSS broke the filter header in both the complex demo
> and the demo.html page.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PredixDev/px-data-table/pull/126#issuecomment-305651085>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA0WTHHS4QJE9K-LIFO7wX8gAMMeqoQQks5r_04WgaJpZM4NPx60>
> .
>

mdwragg commented:
💯 Thanks!

evandana commented:
@mdwragg I have pushed fixes for those issues. Note that there is a right border missing on the table. Is that something new?

mdwragg commented:
Yeah, I just checked it out and that appears to be a regression @evandana.

mdwragg commented:
Sorry, should've made that clearer, the right border disappearance appears to be a regression only in this PR. In master this does not occur.

mdwragg commented:
Hi @evandana - I split the tests up (https://github.com/PredixDev/px-data-table/tree/remote-pr) and copied the changes out to a branch on this repo (so I could run the whole suite of tests against Sauce Labs in our CI environment).

Sadly the tests still fail in IE... We need them working in IE before we can accept the change. I'll leave that branch around for a while so you can see the pattern for splitting the tests up. Thanks!

evandana commented:
Thanks! I split them up myself, but haven't pushed that change. I'll check
your branch. Thanks for your help with it.

I'm a little stuck on how to test with IE since I'm on a mac. What do you
use for testing with IE?

On Wed, Jul 5, 2017 at 4:10 PM Martin Wragg <notifications@github.com>
wrote:

> Hi @evandana <https://github.com/evandana> - I split the tests up (
> https://github.com/PredixDev/px-data-table/tree/remote-pr) and copied the
> changes out to a branch on this repo (so I could run the whole suite of
> tests against Sauce Labs in our CI environment).
>
> Sadly the tests still fail in IE... We need them working in IE before we
> can accept the change. I'll leave that branch around for a while so you can
> see the pattern for splitting the tests up. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PredixDev/px-data-table/pull/126#issuecomment-313212909>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA0WTBbamAFfacIE7CJDjgfUl2KIlzIEks5sK-3JgaJpZM4NPx60>
> .
>

mdwragg commented:
Good point.... We either use the free VMs from https://modern.ie (https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/) or the 'manual test session' option on SauceLabs (there you can pick an IE VM that gives you 10 minutes of use).

To access a local running version of your tests do; `wct -p` from one terminal window, then in another terminal window using ngrok (link below), do `ngrok http 2000` - You'll then get URL to your tests that you can access from anywhere (either a local VM or SauceLabs VM) by going to `http://<ngrok url>/components/px-data-table/generated-index.html`.

https://ngrok.com/

mdwragg commented:
💯 🐐 👍 🥇 

Thanks!
